### PR TITLE
Add step length sampling to Klein-Nishina demo

### PR DESCRIPTION
**Pull request by sethrj**
Add cross sections and sample the mean path length during transport. Tally interaction data, and then collapse after each transport step into a z-dependent energy deposition profile.

### CODE REVIEW

File: app/demo-interactor/KNDemoIO.hh [hunk 1]
```diff
@@ -34,8 +37,8 @@ struct KNDemoResult
     using size_type = celeritas::size_type;
 
     std::vector<double>    time;           //!< Real time per step
-    std::vector<double>    edep;           //!< Energy deposition per step
-    std::vector<size_type> alive;          //!< Num living tracks
+    std::vector<size_type> alive;          //!< Num living tracks per step
+    std::vector<double>    edep; //!< Energy deposition along the grid
```
Comment (pcanal) [inline app/demo-interactor/KNDemoIO.hh:41]:
```suggestion
    std::vector<size_type> alive;          //!< Num living tracks per step
    std::vector<double>    edep;          //!< Energy deposition along the grid
```

File: app/demo-interactor/PhysicsArrayCalculator.hh [hunk 2]
```diff
@@ -0,0 +1,50 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file PhysicsArrayCalculator.hh
```
Comment (pcanal) [inline app/demo-interactor/PhysicsArrayCalculator.hh:6]: Is this class specifically for KNDemo or should be in physics/base?

File: app/demo-interactor/KNDemoIO.hh (hunk 1 above)
Comment (sethrj) [reply app/demo-interactor/KNDemoIO.hh]: Thanks! Must've been left unformatted from the previous merge. I reformatted all the stuff in this directory.

File: app/demo-interactor/PhysicsArrayCalculator.hh (hunk 2 above)
Comment (sethrj) [reply app/demo-interactor/PhysicsArrayCalculator.hh:6]: The idea was to make this class a starting point for a full physics vector/table-type class that would move into physics/base (or physics/em if it's only applicable to the EM models and the geant4 silly cross section scaling). So for expediency I've left it here in the celeritas namespace as a straw-man class. I think once we get a second model/process fully implemented we can start generalizing it. Sound reasonable to you?

**Comment by tmdelellis** [comment]
I have looked over this code quite extensively in the upcoming CPU extension.  I have no objections.

**Review by tmdelellis — APPROVED** [review]

**Pull request merged**